### PR TITLE
Support for Julia 1.0+

### PR DESCRIPTION
**Pull request by gmassman**
# Support for Julia >= 1.0

## Types of changes

This PR implements the following changes:

* [ ] :sparkles: New feature (A non-breaking change which adds functionality).
* [ ] :bug: Bug fix (A non-breaking change, which fixes an issue).
* [X] :boom: Breaking change (fix or feature that would cause existing functionality to change).

## :clipboard: Additional detail
- changes from **@gmassman**
  - Add Project and Manifest toml files with updated dependencies
  - Remove root REQUIRE file
  - Use `Humanize.datasize()` to fix function call error
  - Wrap right hand side of bit shift operator in parenthesis
  - Declare `global function linear_search()` in unit test to fix deprecation

- changes from **@timbitz**
  - Update all deprecated syntax for types and function parameters
  - Re-write LocationIterator syntax (start/done/next) to Base.iterate
  - Replace deprecated base function names (eg. copy!, srand, endof)

## :ballot_box_with_check: Checklist

- [X] :art: The changes implemented is consistent with the [julia style guide](https://docs.julialang.org/en/v1/manual/style-guide/).
- [ ] :blue_book: I have updated and added relevant docstrings, in a manner consistent with the [documentation styleguide](https://docs.julialang.org/en/stable/manual/documentation/).
- [ ] :blue_book: I have added or updated relevant user and developer manuals/documentation in `docs/src/`.
- [X] :ok: There are unit tests that cover the code changes I have made.
- [X] :ok: The unit tests cover my code changes AND they pass.
- [ ] :pencil: I have added an entry to the `[UNRELEASED]` section of the manually curated `CHANGELOG.md` file for this repository.
- [X] :ok: All changes should be compatible with the latest stable version of Julia.
- [ ] :thought_balloon: I have commented liberally for any complex pieces of internal code.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -3,8 +3,7 @@ os:
   - linux
   - osx
 julia:
-  - 0.5
-  - 0.6
+  - 1.0
```
Comment (kescobo) [inline .travis.yml:6]: Can we add tests for latest release (1.3) and nightly as well? Can do

```
matrix:
 allow_failures:
 - julia:
    - 1.3
    - nightly
```

If we don't want to support them, but worth testing at least.

File: test/runtests.jl [hunk 2]
```diff
@@ -1,18 +1,19 @@
...
+using Test
+using Random
 
-srand(12345)
+Random.seed!(12345)
 
 # simple DNA sequence type
 @enum Nuc A C G T
-type DNASeq
+mutable struct DNASeq
```
Comment (kescobo) [inline test/runtests.jl:10]: Looks like this used to be immutable - does it need to be mutable now? Note: if the `data` field is a vector, you can still mutate the underlying vector even if the `struct` wrapping it is technically immutable.

File: test/runtests.jl [hunk 3]
```diff
@@ -258,18 +259,18 @@ end
...
-        while (loc = searchindex(text, Vector{UInt8}(query), loc + 1)) > 0
+        while (loc = first(something(findnext(query, textstr, loc + 1), 0:-1))) > 0
             push!(locs, loc)
         end
         return locs
     end
 
-    text = open(read, joinpath(dirname(@__FILE__), "lorem_ipsum.txt"))
+    text = read(joinpath(dirname(@__FILE__), "lorem_ipsum.txt"))
+    textstr = String(copy(text))
```
Comment (kescobo) [inline test/runtests.jl:272]: This doesn't look like it gets used... maybe meant [for here](https://github.com/BioJulia/FMIndexes.jl/pull/9/files#diff-fce720c43af3c52c862fd7451c7374b8R288)?

File: test/runtests.jl [hunk 4]
```diff
@@ -258,18 +259,18 @@ end
 end
 
 @testset "full-text search" begin
-    function linear_search(query)
+    global function linear_search(query)
```
Comment (kescobo) [inline test/runtests.jl]: Why is this needed?

**Review by kescobo — CHANGES_REQUESTED** [review]
Looks great! Thanks for taking the time to do this. I have a couple of comments/questions

File: test/runtests.jl (hunk 2 above)
Comment (timbitz) [reply test/runtests.jl:10]: In old julia syntax, 'type' is mutable, and 'immutable' is immutable, so I just maintained the original mutability.  But I think you're right that it doesn't need to be mutable, but it's just runtests.jl so i'm not sure this matters either way.

File: test/runtests.jl (hunk 3 above)
Comment (timbitz) [reply test/runtests.jl:272]: lol yeah this isn't doing anything, I think I was testing something and forgot to delete this.

File: .travis.yml (hunk 1 above)
Comment (timbitz) [reply .travis.yml:6]: Yeah this looks good.  Thanks for reviewing!

File: test/runtests.jl (hunk 4 above)
Comment (timbitz) [reply test/runtests.jl]: @gmassman

File: .travis.yml (hunk 1 above)
Comment (gmassman) [reply .travis.yml:6]: I'll add those to the yml, thanks for the review.

File: test/runtests.jl (hunk 3 above)
Comment (gmassman) [reply test/runtests.jl:272]: `textstr` is used for the linear_search. I'm going to pass it into the local linear_search as an arg though and remove the global.

File: test/runtests.jl (hunk 4 above)
Comment (gmassman) [reply test/runtests.jl]: That was fixing a deprecation warning from julia 0.7... I'll remove it and pass in `textstr` so no need for global.

File: test/runtests.jl (hunk 2 above)
Comment (kescobo) [reply test/runtests.jl:10]: Oh right - I thought `type` was the old `struct`, I forgot that `type` was mutable by default.

>but it's just runtests.jl so i'm not sure this matters either way.

This is true. I don't feel strongly about changing it 🤷‍

File: .travis.yml [hunk 5]
```diff
@@ -3,10 +3,18 @@ os:
...
+  - 1.1
+  - 1.2
```
Comment (kescobo) [inline .travis.yml]: Since 1.3 is the current release, and 1.0 is the long-term stable, I don't think we need to test on 1.1 and 1.2. I'm fine to leave them, just makes CI take longer.

File: .travis.yml [hunk 6]
```diff
@@ -3,10 +3,18 @@ os:
...
+  - 1.3
+  - nightly
 notifications:
   email: false
+matrix:
+  allow_failures:
+    - julia:
+      - 1.3
```
Comment (kescobo) [inline .travis.yml]: Is 1.3 actually broken?

Comment (gmassman) [reply .travis.yml]: oh, no it's not. I'll update again

**Comment by kescobo** [comment]
Weird - it looks like on the last build, only nightly failed, but it still said it failed overall 🤔

**Comment by gmassman** [comment]
Yeah, I think maybe the matrix: syntax isn't right? I removed it so maybe it'll work this time.

**Comment by kescobo** [comment]
Hmm... I pulled it from my [repo here](https://github.com/BioJulia/Microbiome.jl/blob/master/.travis.yml#L17-L21), which I think used to work. Then again - that's currently passing on nightly...

**Comment by gmassman** [comment]
Nightly is failing because it doesn't install test dependencies. I'll add `test/Project.toml` file and push again.

**Comment by kescobo** [comment]
LGTM 👍

**Comment by gmassman** [comment]
Cool, thanks again for the review Kevin!

**Comment by kescobo** [comment]
No sweat! I don't have merge privileges though. @BenJWard @bicycle1885  - I think this is good to go.
